lib/util/result.spec.ts

```diff
@@ -1,102 +1,178 @@
+
+      const res5 = await new AsyncResult<string, string>((_, reject) =>
+        reject('oops')
+      ).transform(afn);
+      expect(res5).toEqual(Result.err('oops'));
     });
 
-    it('returns fallback value for error result', () => {
-      const err = new Error('oops');
-      const res = Result.err(err);
-      expect(res.catch(42)).toBe(42);
+    it('logs and returns error for transform failure', () => {
+      const res = Result.ok('foo').transform(() => {
+        throw 'oops';
+      });
+      expect(res).toEqual(Result.err('oops'));
+      expect(logger.logger.warn).toHaveBeenCalledWith(
+        expect.anything(),
+        'Result: unhandled transform error'
+      );
     });
-  });
 
-  describe('value', () => {
-    it('returns successful value', () => {
-      const res = Result.ok(42);
-      expect(res.value).toBe(42);
+    it('logs and returns error for async transform failure', async () => {
+      const res = await Result.ok('foo').transform(() =>
+        Promise.reject('oops')
+      );
+      expect(res).toEqual(Result.err('oops'));
+      expect(logger.logger.warn).toHaveBeenCalledWith(
+        expect.anything(),
+        'Result: unhandled async transform error'
+      );
+    });
+
+    it('transforms successful promises', async () => {
+      const res = await Result.wrap(Promise.resolve('foo')).transform(fn);
+      expect(res).toEqual(Result.ok('FOO'));
     });
 
-    it('returns undefined value for error result', () => {
-      const err = new Error('oops');
-      const res = Result.err(err);
-      expect(res.value).toBeUndefined();
+    it('bypasses transform of failed promises', async () => {
+      const res = await Result.wrap(Promise.reject('bar')).transform(fn);
+      expect(res).toEqual(Result.err('bar'));
+    });
+
+    it('handles failed transform of successful promise', async () => {
+      const res = await Result.wrap(Promise.resolve('foo')).transform(() => {
+        throw 'bar';
+      });
+      expect(res).toEqual(Result.err('bar'));
+      expect(logger.logger.warn).toHaveBeenCalledWith(
+        expect.anything(),
+        'AsyncResult: unhandled transform error'
+      );
+    });
+
+    it('handles failed async transform of successful promise', async () => {
+      const res = await Result.wrap(Promise.resolve('foo')).transform(() =>
+        Promise.reject('bar')
+      );
+      expect(res).toEqual(Result.err('bar'));
+      expect(logger.logger.warn).toHaveBeenCalledWith(
+        expect.anything(),
+        'AsyncResult: unhandled async transform error'
+      );
+    });
+
+    it('handles chained failure', async () => {
+      const fn1 = (x: string): Result<string, string> =>
+        Result.ok(x.toUpperCase());
+      const fn2 = (x: string): Result<string[], number> =>
+        Result.ok(x.split(''));
+      const fn3 = (x: string[]): Result<string, boolean> =>
+        Result.ok(x.join(' '));
+
+      const res: Result<string, string | number | boolean> = await Result.wrap<
```

Comment:
This is illustration of how error types are aggregating via the chain of transforms.

We want to know which errors can be produced in a particular code fragment, but Typescript doesn't support it very well. Because of that, I decided to aggregate all the errors, but log warnings every time some uncaught error was thrown. This way, `Result` type will reflect the errors we catch and handle explicitly. Hopefully, this approach is good enough given the tools we have.